### PR TITLE
Stop requestslib from overriding auth headers with data from .netrc file

### LINKED ISSUE #337: Esoteric Bug with regards to authentication via `requests` `headers=` argument being overwritten by `.netrc`
### Describe the bug
It's an odd one that set me back around a day: I hope to save people a similar amount of frustration.

When running a python model on an all purpose cluster, I would get this error:
```
10:54:28    Error getting status of cluster.
10:54:28     b'{"error_code": "403", "message": "Invalid access token."}'
```
**despite putting a brand new token into the `profiles.yml` for the target corresponding all purpose cluster.**

Turns out that the my `.netrc` file was to blame.  It had an expired token for the same `host:`.  Also turns out that the `requests` library overrides authentication methods passed through `header=` argument with the contents of the `.netrc` file, as stated here: https://requests.readthedocs.io/en/latest/user/authentication/#netrc-authentication

To be 100% clear, my `.netrc` file looked like this (replacing sensitive info below):

`.netrc`:
```
machine <my-workspace>.cloud.databricks.com
login token
password <expired_token>
```

My `.profiles` contained a target like:
```
all_purpose:
  type: databricks
  host: <my-workspace>.cloud.databricks.com
  http_path: <sql_path_for_all_purpose_cluster>
  catalog: <catalog>
  schema: <schema>
  token: <valid_token>
  threads: 1
```

Then the weird part: 
 * `requests.py` does pass the `<valid_token>` in to the headers argument in the `get_cluster_status` call.
 * With that said, `requests.py` would give me an error from this method because it defaulted to my `.netrc` token: https://github.com/databricks/dbt-databricks/blob/main/dbt/adapters/databricks/python_submissions.py#L245

### Steps To Reproduce

Fairly easy to trigger:

 * Put an **expired or invalid** token into the `.netrc` file.  
   * Note that using a `.netrc` file is encouraged in the [Databricks API documentation](https://docs.databricks.com/dev-tools/api/latest/authentication.html#store-tokens-in-a-netrc-file-and-use-them-in-curl)
 * Put a **valid** token in the `~/.dbt/profiles.yml` for an all purpose cluster that we want to use to run a python model.
 * Run the model, get the error from here: https://github.com/databricks/dbt-databricks/blob/main/dbt/adapters/databricks/python_submissions.py#L255

### Expected behavior

I'd expect the token provided in my `profiles.yml` to take precedence.  

### Screenshots and log output

Not needed IMO, already know the cause.

### System information

Relevant versions are:
 * `dbt-databricks==1.5.0`
 * `requests==2.28.1`

### Additional context

I would say the solution is to ensure the `dbt-databricks` code uses the `auth` argument instead.  It seems a "bearer token" auth method doesn't come included in `requests` (which seems odd but okay).  To this end there is a good SO solution to this: https://stackoverflow.com/a/58055668

I would also highlight that while this feels like it might only be a *me* problem, Databricks has lots of docs around using `.netrc` file, as linked before: [Databricks API documentation](https://docs.databricks.com/dev-tools/api/latest/authentication.html#store-tokens-in-a-netrc-file-and-use-them-in-curl).  Doesn't seem too far-fetched someone else will fall into this trap, and its a hard one to get out of.

### PR DESCRIPTION
**Pull request by susodapop**
Resolves #337

### Description

This pull request refactors the credentials handling in `python_submissions.py` using the workaround suggested in #337. The desired behaviour from this change is not obvious when looking only at the code diff. However the idea here is to explicitly order Python requests to use Bearer authentication with an explicitly provided token. If this is not done explicitly, there is an edge case where a local `.netrc` file can override an auth token present in `profiles.yml`.

There's no automated test for this because I'm not sure how to deposit a .netrc file in a test environment.

I reproduced the issue manually by adding this to `~/.netrc` on my workstation:

```
machine <my-workspace>.cloud.databricks.com
login token
password <expired_token>
```

I removed the skip marker on `test_python_uc_databricks_sql_endpoint` and ran this integration test on the `main` branch. The test failed with this exception:

```
ERROR    stdout_log:eventmgr.py:66 [0m18:58:04     b'{"error_code":"403","message":"Invalid access token."}'
ERROR    file_log:eventmgr.py:66 [0m13:58:04.325701 [error] [MainThread]:    b'{"error_code":"403","message":"Invalid access token."}'
```

Then I applied the change in this pull request and re-ran the test. It passed.

### Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

**Review by andrefurlan-db — CHANGES_REQUESTED** [review]
it needs to work with oauth as well. Token is not passed in config

**Comment by susodapop** [comment]
I've updated the code to use the `HeaderFactory` provided by `databricks.sdk` and verified it still works with PAT auth. Need to do a quick test with OAuth and maybe basic auth to make sure those cases are also handled.

**Comment by susodapop** [comment]
I found that the same issue we're fixing here (.netrc overrides auth headers in requests made by dbt-databricks) also affects our databricks-sdk-py repo (.netrc overrides auth headers in requests made for oauth handshake).

The fix in this PR will fix dbt-databricks .netrc woes if the auth method is PAT. But it will only fix the issue for oauth use-cases after databricks-sdk-py is updated with this patch https://github.com/databricks/databricks-sdk-py/issues/107 and we update the dbt-databricks dependency to incorporate that fix.

Note: I found the same bug in `databricks-sql-connector`'s custom implementation of OAuth handshakes _but this doesn't affect dbt-databricks because the dbt adapter uses the OAuth implementation in the SDK_.

**Comment by susodapop** [comment]
Now that databricks-sdk has pushed the parallel fix for oauth I've rebased on `main`. This included a refactor of `SessionCredentials` base on a breaking change in databricks-sdk.

**Comment by susodapop** [comment]
Looks like there is now a parallel PR that bumps the databricks-sdk version and includes the `RefreshableCredentials` refactor: https://github.com/databricks/dbt-databricks/pull/349

**Review by andrefurlan-db — DISMISSED** [review]
Left one comment that can be addressed in a separate PR

**Review by andrefurlan-db — COMMENTED** [review]
now that I think about it, is there a chance of the token expiring while polling and executing?

Maybe it is better to not set self.auth and get the headers before each request from self._credential_provider ?

**Comment by benc-db** [comment]
@susodapop what is this status of this PR?  Also, if we are going to bring this in, we really do need to make sure we're sending the right user agent, since all of the tracking for this library depends on it.

**Comment by susodapop** [comment]
It was deprioritised at the time because were chasing down OOM issues in dbt-databricks. We should get this mergeable and release it, although the bug radius is pretty small since it requires `.netrc` to be present and Python models to be in use.

**Comment by lennartkats-db** [comment]
@susodapop @andrefurlan-db any update on this? I just ran into this issue as well. This issue should be fairly common among Databricks customers since we still have lots of docs and community content calling out that `.netrc` should be used. #337 causes materialized views to fail with `Error getting info for materialized view/streaming table: <catalog.schema.table>`

**Comment by benc-db** [comment]
Looks like there is a conflict to resolve.

File: dbt/adapters/databricks/python_submissions.py [hunk 1]
```diff
@@ -31,11 +47,12 @@ def __init__(self, parsed_model: Dict, credentials: DatabricksCredentials) -> No
         self.timeout = self.get_timeout()
         self.polling_interval = DEFAULT_POLLING_INTERVAL
         self.check_credentials()
-        self.auth_header = {
-            "Authorization": f"Bearer {self.credentials.token}",
+        self.extra_headers = {
             "User-Agent": f"dbt-labs-dbt-spark/{DBT_SPARK_VERSION} (Databricks)",
```
[the remark this answers is not in the document]
Comment (benc-db) [reply dbt/adapters/databricks/python_submissions.py:52]: This looks like the wrong user agent @susodapop lol.

**Comment by susodapop** [comment]
My fault. I failed to pull the latest `main` before synchronising. Will fix locally and push again.

File: dbt/adapters/databricks/python_submissions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (susodapop) [reply dbt/adapters/databricks/python_submissions.py]: Fixed!

**Review by benc-db — DISMISSED** [review]

**Comment by susodapop** [comment]
Alright I've pushed all the changes from code review. I'm going to manually test this with a bad `.netrc` locally and wait for the integration tests to pass before requesting a merge.

**Comment by susodapop** [comment]
I have verified that the python models tests fail without my change and a bad `.netrc` is present. With my fix applied, the python model tests work.

**Comment by susodapop** [comment]
Alright, I've pushed the final version of this code. I moved the `BearerAuth` class into `connections.py` because we're about to use it there to fix an issue with streaming tables / materialized views. 

I also simplified the code by setting the `Session.auth` property rather than setting the `.auth` property for every single network request.

**Review by benc-db — APPROVED** [review]
